photon-api/src/main/scala/com/linkedin/photon/ml/projector/IndexMapProjectorRDD.scala

```diff
@@ -155,16 +156,51 @@ protected[ml] class IndexMapProjectorRDD private (indexMapProjectorRDD: RDD[(Str
 }
 
 object IndexMapProjectorRDD {
+
   /**
    * Generate index map based RDD projectors.
    *
    * @param randomEffectDataSet The input random effect data set
    * @return The generated index map based RDD projectors
    */
-  protected[ml] def buildIndexMapProjector(randomEffectDataSet: RandomEffectDataSet): IndexMapProjectorRDD = {
-    val indexMapProjectors = randomEffectDataSet.activeData.mapValues(localDataSet =>
-      IndexMapProjector.buildIndexMapProjector(localDataSet.dataPoints.map(_._2.features))
-    )
+  protected[ml] def buildIndexMapProjector(
+      randomEffectDataSet: RandomEffectDataSet): IndexMapProjectorRDD = {
+
+    val originalSpaceDimension = randomEffectDataSet
+      .activeData
+      .map { case (_, ds) => ds.dataPoints.head._2.features.length }
+      .take(1)(0)
```

Comment:
Wouldn't it make more sense to take 1, then get the dimension from that instead of mapping over all data and creating a new RDD then taking the first data point? Is Spark with RDDs instead of DataFrame/DataSet smart enough to only lazy evaluate one record?

Reply:
Yes, it's lazy, so this only actually gets executed for one record.